software/tests/baremetal/fence/main.c

```diff
@@ -0,0 +1,156 @@
+// Copyright 2021 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Author: Samuel Riedel, ETH Zurich
+
+#include <stdint.h>
+#include <string.h>
+
+#include "data.h"
+#include "dma.h"
+#include "encoding.h"
+#include "mempool_dma_frontend.h"
+#include "printf.h"
+#include "runtime.h"
+#include "synchronization.h"
+
+// Size in words
+#define SIZE (12288)
+
+uint32_t l1_data[SIZE] __attribute__((section(".l1_prio")));
+uint32_t volatile l2_var[2] __attribute__((section(".l2")));
+uint32_t volatile error_no_fence __attribute__((section(".l1")));
+uint32_t volatile error_fence __attribute__((section(".l1")));
+
+dump(no_fence, 10);
+dump(fence, 11);
+
+// Simple Litmus test
+//
+// location l0 and l1 are initialized to `golden` and after reading overwritten
+// with `overwrite`
+//
+// b = `overwrite`
+// Core 0    | Core 1
+// lw a (l1) | lw a (l0)
+// fence     | fence
+// sw b (l0) | sw b (l1)
+//
+// With the fence, at least one core must have a=`golden`, but it is possible
+// that one core has a=`overwrite`. Without a fence, it is allowed that both
+// cores end up with a=`overwrite`. This test first executes without a fence to
+// trigger this "inconsistency" and reports on whether the scenario is possible
+// to validate that the test could actually fail. The second iteration runs with
+// a fence and checks for the one prohibited outcome (both cores have
+// a=`overwrite`). To maximize the chance of writes overtaking reads, we use the
+// DMA to stress the read channel into the L2 memory.
+//
+// Excepted outcome:
+// Without a fence:
+// [DUMP] Core 0: 0x00a = 5678  // a = `overwrite`. Inconsistency triggered
+// [DUMP] Core 1: 0x00a = 5678  // a = `overwrite`. Inconsistency triggered
+// With the fence
+// [DUMP] Core 0: 0x00b = 1234 // a = `golden`. One of them being `overwrite`
+// [DUMP] Core 1: 0x00b = 1234 // a = `golden`. would also be allowed.
```

Comment:
Put to newline:
`// (One of them being `overwrite` would also be allowed)`

Reply:
We do an `and` operation on the error. So the final return value is the logic `and` of whether core 0 and core 1 got `overwrite` as their read value.